### DISCUSSION TITLE
GitLens interactive rebase feature & rebase --edit-todo

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#2976] [General]
I'm not sure if this is the right/best place to post this, so please forgive me if not.

I'm working on expanding the capabilities of the interactive rebase feature I shipped recently in GitLens to continue through the rebase as it pauses.

See here for the current WIP:
![rebase](https://user-images.githubusercontent.com/641685/104369783-fcc8d500-54eb-11eb-9209-ee34f9756bef.gif)

My question is since I'm currently allowing the user to modify the rebase steps that are still "todo", do I need to call `git rebase --edit-todo`? It seems to work fine without it, but I'm not sure if I'm getting lucky or causing not yet seen issues.

Thanks!

**Comment by derrickstolee** [comment]
My guess is that `git rebase --edit-todo` is the most future-proof way to do it. Without that, you are relying on the "internal" data layout of Git and `git rebase` in particular.

While it is very unlikely that we would change the name of the TODO file, it is certainly a possibility.

@dscho will know more about whether current work in `git rebase` has potential to disrupt your work.

**Comment by eamodio** [comment]
Since Git (AFAIK) doesn't provide any access to the rebase steps other than the TODO file, I'm highly reliant on the TODO file and its structure (even before a rebase starts). Although, I guess I also don't understand the usage of `--edit-todo` because as documented it doesn't take any other input -- so it would rely on some other mechanism for updating the TODO file or something right?

Also, FYI, this feature is built on the TODO file itself, as I structured this as a "custom editor" in VS Code, which means this is just a UI "view" on top of the actual file itself.

I also currently rely on a lot of other rebase files to determine the current state:

https://github.com/eamodio/vscode-gitlens/blob/3509a3a247ee7b799c92334874088dd7e8764149/src/git/gitService.ts#L2447-L2515

I'm not sure if there is any other way for Git to provide me with that information (without reading all those files directly).

**Reply by siprbaum** [reply]
@eamodio The link you posted resuls in a 404 here

**Reply by eamodio** [reply]
Updated the link -- thanks!

**Reply by dscho** [reply]
In Git's test suite, we override `GIT_SEQUENCE_EDITOR` to edit the `git-rebase-todo` file.

As @derrickstolee hinted, there are good reasons to use `git rebase --edit-todo` to edit this file: it lints the written todo list (and exits with a non-zero code if an error was detected), and it also expands the commit hashes to avoid collisions.

**Reply by eamodio** [reply]
Ah! Awesome -- that makes a lot of sense and tells me how I should use it more! Thank you both!

**Reply by eamodio** [reply]
@dscho It seems that if there is an issue in the TODO file, `rebase --edit-todo` reports the errors, but then opens the TODO file for editing. Is there any way to avoid that part? I just want the linting & fixup.

**Reply by dscho** [reply]
You could set the editor to `cat`, for example, or to `true`, e.g. `git -c sequence.editor=true rebase --edit-todo`.

**Reply by eamodio** [reply]
Oh sweet! Thanks!
